### PR TITLE
Implement Nexus Serializer for Temporal Payloads

### PR DESCRIPTION
**Pull request by bergundy**
**What changed?**

See description

**Why?**

The server should do best effort conversion between Nexus Content and Temporal Payloads so pure Temporal applications don't have to deal with the separate Nexus Content concept.

**How did you test it?**

Added tests.

**Potential risks**

Non-standard content is converter into a Payload with metadata encoding of `unknown/nexus-content`. Users will need to prepare for that or their Nexus requests may fail.
I believe this is the best option we have to avoid lossy conversion and expose as much information and empower the application.
I considered using `binary/plain` for non-standard content but I'd rather not have these payloads accidentally slip into user's applications.

**Is hotfix candidate?**

Far from it.

### CODE REVIEW

File: common/nexus/payload_serializer.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer.go:154]: You may as well just have `errSerializer` be a string prefix; it doesn't look like we gain anything by wrapping it into the error tree

File: common/nexus/payload_serializer.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer.go:190]: Can you report back the invalid encoding so its obvious for users?

File: common/nexus/payload_serializer_test.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer_test.go:38]: take a `*testing.T` in and you can just use `require.NoError` so we get a better test output on failure

File: common/nexus/payload_serializer_test.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer_test.go:118]: Nit: This is the zero value and needn't be specified

File: common/nexus/payload_serializer_test.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer_test.go:125]: Nit: zero values needn't be specified

File: common/nexus/payload_serializer_test.go
Comment (tdeebswihart) [inline common/nexus/payload_serializer_test.go:48]: Most of these tests have neither an expected error nor an expected output, what are we actually testing for in those cases? Are we only testing to for "no error and a zero-valued output"?

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:36]: It's in the `nexus` package, so that will show up as part of the error identifier (although not in the error message) but I think that's fine.

File: common/nexus/payload_serializer.go
Comment (bergundy) [inline common/nexus/payload_serializer.go:34]: I'm going to rename this to `payloadSerializer` and avoid the stuttering.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:135]: No, it's just our old convention where we didn't specify the message type.
The message type is only required for languages that don't have runtime types (JS 👀).

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:141]: Yes, there's a separate Nexus Go SDK that expects the serializer to implement this specific interface.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:154]: Mostly did this to satisfy our linter.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:190]: Yeah, good point.
I originally had a panic here because it's supposed to be unreachable code but changed to an error to be nicer.

File: common/nexus/payload_serializer_test.go
Comment (bergundy) [reply common/nexus/payload_serializer_test.go:38]: I think that's overkill, I only use it this helper to serialize things I know are serializable but I don't mind applying your suggestion.

File: common/nexus/payload_serializer_test.go
Comment (bergundy) [reply common/nexus/payload_serializer_test.go:48]: Yes, the error cases are around marshaling or invalid type specified.
I could test the latter, just didn't feel like it was really adding much.
Marshaling shouldn't fail.

File: common/nexus/payload_serializer_test.go
Comment (bergundy) [reply common/nexus/payload_serializer_test.go:118]: `nexus.Header{}` is a `map[string]string`, the zero value for a map is nil.

File: common/nexus/payload_serializer_test.go
Comment (bergundy) [reply common/nexus/payload_serializer_test.go:125]: Same, zero value is `[]byte(nil)` which isn't the same as `[]byte{}`.

File: common/nexus/payload_serializer_test.go
Comment (bergundy) [reply common/nexus/payload_serializer_test.go:48]: FTR, I checked the coverage and I'm fairly satisfied with it.

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:176]: I don't think any SDKs leave this off anymore and haven't for a while, I think we can require usage of a somewhat modern SDK version for use with Nexus protos

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:152]: This is confusing too, can you just make this stuff part of the switch below? You can set data in each arm of the switch, there's not that much code reuse here. Just handle each type specifically. I can suggest exact rewrite if it helps.

**Comment by cretz** [comment]
(I don't usually do reviews in this repository but I was asked in this case)

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:155]: Again, I could make a linter exception here. I was just "following the rules".

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:176]: I'm fine with that.

**Review by bergundy — COMMENTED** [review]
Thanks for the reviews.

File: common/nexus/payload_serializer.go
Comment (cretz) [reply common/nexus/payload_serializer.go:34]: Would be willing to wait until then to review since how it is used affects how it is written

File: common/nexus/payload_serializer.go
Comment (cretz) [reply common/nexus/payload_serializer.go:80]: Yeah, more and more the single param sounds good.

File: common/nexus/payload_serializer.go
Comment (cretz) [reply common/nexus/payload_serializer.go:113]: An absent content type + data is always asymmetric because it's more loosely defined inbound but we want to explicitly set content type outbound when data is present. So discussion of absent content-type + data only applies in one direction.

**Comment by bergundy** [comment]
Merged the detection and serialization logic and address the rest of the comments.
PTAL.

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:55]: Why remove these here when you can just filter and not set them in the one place you actually copy headers, the `setUnknownNexusContent` call?

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:57]: Why do you care what the rest of the headers are so long as the content type is correct? I think you're overthinking here and can just remove this conditional

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:67]: I still think this should be `binary/plain`

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:40]: To me this reads clearer as a future maintainer, but don't have to accept it (just typed real quick, so probably some typos):

```go
// Deserialize implements nexus.Serializer.
func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
	payloadRef, ok := v.(**commonpb.Payload)
	if !ok {
		return fmt.Errorf("cannot deserialize into %v", v)
	}
	payload := &commonpb.Payload{Metadata: map[string][]byte{}, Data: content.Data}
	*payloadRef = payload

	// Absent content type with no data means null, with data is just binary
	if contentType := content.Header["type"]; contentType == "" {
		if len(content.Data) == 0 {
			payload.Metadata["encoding"] = []byte("binary/null")
		} else {
			payload.Metadata["encoding"] = []byte("binary/plain")
		}
		return nil
	}

	// Different content types make different payloads
	if mediaType, params, err := mime.ParseMediaType(contentType); err == nil {
		switch mediaType {
		case "application/x-temporal-payload":
			if err := payload.Unmarshal(content.Data); err != nil {
				return fmt.Errorf("failed to unmarshal payload: %w", err)
			}
			return nil
		case "application/json":
			if protoMessageType := params["protobuf-message-type"]; protoMessageType != "" {
				payload.Metadata["encoding"] = []byte("json/protobuf")
				payload.Metadata["messageType"] = []byte(protoMessageType)
			} else {
				payload.Metadata["encoding"] = []byte("json/plain")
			}
			return nil
		case "application/x-protobuf":
			if protoMessageType := params["protobuf-message-type"]; protoMessageType != "" {
				payload.Metadata["encoding"] = []byte("binary/protobuf")
				payload.Metadata["messageType"] = []byte(protoMessageType)
				return nil
			}
			return fmt.Errorf("protobuf message type required for application/x-protobuf content type")
		case "application/octet-stream":
			payload.Metadata["encoding"] = []byte("binary/plain")
			return nil
		}
	}

	// Unknown or bad content type
	payload.Metadata["encoding"] = "binary/unknown-nexus"
	for k, v := range content.Header {
		// Don't want length or encoding
		if k != "length" && k != "encoding" {
			payload.Metadata[k] = []byte(v)
		}
	}
	return nil
}
```

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:127]:
```suggestion
		return &nexus.Content{}, nil
```

A nil map is just fine here

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:146]: I think, like with the code in deserialize, you're being overly strict on the metadata. If it's our encoding, we can choose to ignore other metadata

File: common/nexus/payload_serializer.go
Comment (cretz) [inline common/nexus/payload_serializer.go:87]: I think you're being overly strict on the mime type params. If it's our content type, we can ignore params we don't care about

**Review by cretz — APPROVED** [review]

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:57]: I don't want to lose information in the conversion

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:67]: This could also be content header that's not `type`, and I'd rather not lose this information.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:87]: again, I'd rather not lose this information. Either you're sticking to the standard or you're a custom nexus payload.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:127]: I guess I can't require serializer implementations to provide non nil headers. I'll apply your suggestion.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:146]: I don't think you're right here.
If a user customizes serialization to add more information, we should treat their payload as opaque and non standard and avoid losing metadata during the conversion.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:55]: This is mostly to simplify the logic below and sanitize any irrelevant fields.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:43]: I personally prefer the prefixed version.

File: common/nexus/payload_serializer.go
Comment (bergundy) [reply common/nexus/payload_serializer.go:40]: I prefer my stricter non-lossy version.

Comment (bergundy) [reply common/nexus/payload_serializer.go:40]: Thanks for the suggestion though.

**Comment by bergundy** [comment]
@cretz I thought about it some more and it may be desirable to be a bit less strict with the conversion but I would like to do that in a way that no information is lost.

In any case, I'll get back to this at some point before merging `nexus` into `main` but I'd rather make some more progress before that.

**Pull request merged**
